doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,318 @@
+  that does not emit redundant data (reusable reclocking).
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+- Accomplishing 'subsource demuxing' using the concept of *subsources* in-use today.
+  The goals of https://github.com/MaterializeInc/materialize/issues/24843 will be
+  accomplished during the process of replacing *subsources* with *tables* (see 
+  solution proposal below).
+  
+- Dealing with Webhook sources. These are fundamentally different than the other
+  storage-owned source types whose operators implement the `SourceRender` trait.
+  While they may also be conveted to 'write to tables' in the future, that work
+  is not in-scope for this proposal.
+
+
+## Solution Proposal
+
+The concept of a *source* will be unified around a single relation representing
+the *progress* of that source's ingestion pipeline. A query of any *source*
+object (of any type) will return the data that is currently in the
+`<source>_progress` collection.
+
+We have been investing in the user experience around *tables*, such that
+replacing *subsources* with *tables* would reduce the overall
+mental overhead for new users.
+
+By allowing tables to be marked as *read-only*, we can easily migrate
+existing subsources to be tables without needing to figure out how to allow
+`INSERT/UPDATE/DELETE` statements for them merge with the data
+written by their corresponding source.
+
+The default `CREATE SOURCE ..` statement will just create a top-level source object
+that represents a single ingestion of data from an upstream system. To actually
+ingest data from that upstream system into a Materialize collection, the user will
+use something akin to
+`CREATE TABLE <table> (<cols>) OF SOURCE <source> WITH (<upstream_ref>)`.
+
+We will allow more than one table to reference the same `<upstream_ref>`, using a
+potentially different set of columns. This allows a user to handle an upstream
+schema change by creating a new table for the same upstream table and then performing
+a blue-green swap operation to switch their downstream dependencies to the new table,
+and then drop the old one.
+
+The existing options to `CREATE SOURCE` for Postgres & MySQL sources that specify
+which tables to ingest (`FOR ALL TABLES`, `FOR SCHEMAS` and `FOR TABLES`) will be
+removed. Instead, all upstream tables to ingest from this source will need to be
+explicitly referenced in a `CREATE TABLE` statement. While this may seem controversial,
+more often than not these options cause upstream data that does not need to be
+ingested into Materialize to be so, and by using explicit statements for each
+table to be ingested it makes Materialize configuration much more amenable to
+object<->state mappings in tools like DBT and Terraform.
```

Comment:
SGTM. We ripped `FOR ALL TABLES` and `FOR SCHEMAS` out of Terraform not too long ago because they didn't play well with Terraform state management. So I don't think removing those features will be terribly painful in practice. 

Honestly, what we might want instead is a SQL command that returns all known upstream tables and columns for a source. That way the console could provide a nice interface for adding tables to a source. (Imagine checking off boxes next to each table/column you wanted to import.)

Reply:
> Honestly, what we might want instead is a SQL command that returns all known upstream tables and columns for a source. That way the console could provide a nice interface for adding tables to a source. (Imagine checking off boxes next to each table/column you wanted to import.)

💯

Reply:
@rjobanp: can we include this functionality in the scope of the epic? Being able to grab the list of tables, columns (+ types) from upstream is a huge usability unlock that I'm afraid will otherwise make it really frustrating to use this new interface. Currently thinking through all the implications of this refactor, I'll have something more detailed to share later in the week.

Reply:
Yes sounds good! Added to the epic here https://github.com/MaterializeInc/materialize/issues/28697